### ISSUE TITLE
Use nanoserde?

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#7758]
https://github.com/not-fl3/nanoserde now suports enums (it didn't before I would think?), it'd be great to try that for lsp-types.

**Comment by lnicola** [comment]
We need to keep in mind that `lsp-types` uses `serde`.

**Comment by lnicola** [comment]
We still have a couple of dependencies pulling in `serde`, though (e.g. `cargo-platform`, `tracing-subscriber`, `semver`).

**Comment by matklad** [comment]
the main problem with serde is not serde per-se, but the amount of code
generated by proc macros. It's not the serde which is slow to compile, but
lsp-types

**Comment by SomeoneToIgnore** [comment]
Looks like this one might not be implemented without some preparations or compromises:

* in some crates, there's `smol_str` that needs regular `serde` and does not support `nanoserde`
* many crates use `deserializer.disable_recursion_limit();` with `serde`, `nanoserde` does not seem to such method in its deserializer, not sure about the implementation details though.

The bullets combined exclude all crates but `rust-analyzer` one.
But that uses `serde_path_to_error` which requires `serde::de::Deserializer`.

**Comment by matklad** [comment]
That's true, but, at the same time, the main issue for us is LSP types. What I had in mind is to first wait until LSP actually publishes machine readable spec, and then using that to generate `nano_lsp_types`. In this case, we would control everything, so ecosystem -wide concerns won't actually be concerns. But machine readable LSP isn't really coming :(

**Comment by SomeoneToIgnore** [comment]
https://github.com/gluon-lang/lsp-types/blob/3cc4b14fe33fc895a999091b9adc4ef10714d485/Cargo.toml#L21-L22

`lsp_types` seem to share the same "adjust or compromise" issue, depending on [`serde_repr`](https://github.com/dtolnay/serde-repr) and `url` with `serde` feature.

A quick look shows that at least `lsp_types::CompletionItem` depends on the enum that uses `serde-repr` to serialize itself, so we need to nanoserdisize those dependencies first.

**Comment by matklad** [comment]
>and url with serde feature.

Not depending on `url` is one of positive side-effects I see from the transition. I think all we need is `pub struct Url(pub String)`, servo's url is a huge source of accidental complexity for us :) 

No quick answer for serde_repr -- my gut feeling is that it shoudl be possible to just directly express what's in the LSP, but this might not work out

**Comment by kjeremy** [comment]
There was some talk over the last week or so of machine readable lsp in one
of the issues.
